### PR TITLE
Transfer learning (SWBD -> WSJ)

### PR DESCRIPTION
**Pull request by pegahgh**

### CODE REVIEW

**Comment by pegahgh** [comment]
@danpovey 
This PR is about transferring information from already-trained model. 
The results are added to training scripts.
Any idea to improve results??

**Comment by danpovey** [comment]
thanks!  I'll try to get to it soon.

**Comment by danpovey** [comment]
Actually don't bother updating the WSJ scripts just yet.
I am reorganizing the WSJ scripts and will update the nnet3/chain scripts to use speed perturbation.

**Comment by jtrmal** [comment]
merged the master (i.e. Dogan's changes) and restarted the build -- it has
failed before because of the timeout.
y.

**Comment by danpovey** [comment]
I have finished with my changes in the WSJ setup so you can go ahead with this.

File: egs/wsj/s5/steps/nnet3/chain/train.py [hunk 1]
```diff
@@ -311,13 +316,23 @@ def train(args, run_opts, background_process_handler):
...
-        logger.info("Initializing a basic network for estimating "
-                    "preconditioning matrix")
-        common_lib.run_kaldi_command(
-            """{command} {dir}/log/nnet_init.log \
-                    nnet3-init --srand=-2 {dir}/configs/init.config \
-                    {dir}/init.raw""".format(command=run_opts.command,
-                                             dir=args.dir))
+        if args.init_raw_model is None:
+            logger.info("Initializing a basic network for estimating "
+                      "epreconditioning matrix")
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/chain/train.py]: typo still here, just different.

File: egs/wsj/s5/steps/nnet3/xconfig_to_configs.py [hunk 2]
```diff
@@ -21,6 +21,12 @@ def get_args():
         description="Reads an xconfig file and creates config files "
                     "for neural net creation and training",
         epilog='Search egs/*/*/local/{nnet3,chain}/*sh for examples')
+    parser.add_argument('--aux-xconfig-file',
+                        help='auxilary Filename of input xconfig file,'
+                        'where its layers can be used as auxilary information'
+                        'used to generated new config file.'
+                        'e.g. generating another newtork using nodes from'
```
Comment (danpovey) [inline egs/wsj/s5/steps/nnet3/xconfig_to_configs.py]: the typo is still there.

File: egs/wsj/s5/steps/nnet3/decode.sh [hunk 3]
```diff
@@ -68,7 +68,7 @@ model=$srcdir/$iter.mdl
 extra_files=
 if [ ! -z "$online_ivector_dir" ]; then
   steps/nnet2/check_ivectors_compatible.sh $srcdir $online_ivector_dir || exit 1
-  extra_files="$srcdir/final.ie.id $online_ivector_dir/ivector_online.scp $online_ivector_dir/ivector_period"
+  extra_files="$online_ivector_dir/ivector_online.scp $online_ivector_dir/ivector_period"
```
Comment (jtrmal) [inline egs/wsj/s5/steps/nnet3/decode.sh]: I guess you're right. Will try to address this today.

Comment (jtrmal) [inline egs/wsj/s5/steps/nnet3/decode.sh]: OK, seems either someone has already fixed it or I;m not sure what's going on. But the decode.sh in master certainly does not have that issue.

**Comment by danpovey** [comment]
@pegahgh, please notice that this now has conflicts!  Let's try to get this checked in fairly soon.
I think it's better if you upgrade this PR to be against kaldi_52 in any case, or otherwise we'll likely have conflicts when merging master into kaldi_52.

**Comment by pegahgh** [comment]
sure Dan, I will prepare the PR ASAP.

**Comment by pegahgh** [comment]
@danpovey 
librispeech->WSJ looks fine, but swbd->AMI still suffer from sampling-rate mismatch!
WSJ->RM should be fine(It seems we have chain script for both sets)! (WSJ->RM is nice because they are both quite small).
It seems there is no nnet3+xent setup for RM, but there is nnet3+chain setup, which is not as good nnet2+xent(2.71% vs. 2.31%)

File: src/nnet3/nnet-component-itf.h [hunk 4]
```diff
@@ -407,6 +407,11 @@ class UpdatableComponent: public Component {
   /// a different value than x will returned.
   BaseFloat LearningRate() const { return learning_rate_; }
 
+  // Sets the learning rate factors to set to this value.
```
Comment (danpovey) [inline src/nnet3/nnet-component-itf.h]: to set to -> to

File: src/nnet3/nnet-utils.cc [hunk 5]
```diff
@@ -579,6 +579,32 @@ void ReadEditConfig(std::istream &edit_config_is, Nnet *nnet) {
         }
       }
       KALDI_LOG << "Set learning rates for " << num_learning_rates_set << " nodes.";
+    } else if (directive == "set-learning-rate-factor") {
+      std::string name_pattern = "*";
+      // name_pattern defaults to '*' if non is given.
```
Comment (danpovey) [inline src/nnet3/nnet-utils.cc]: non -> none

File: src/nnet3/nnet-utils.cc [hunk 6]
```diff
@@ -579,6 +579,32 @@ void ReadEditConfig(std::istream &edit_config_is, Nnet *nnet) {
...
+      config_line.GetValue("name", &name_pattern);
+      BaseFloat learning_rate_factor = -1;
+      if (!config_line.GetValue("learning-rate-factor", &learning_rate_factor)) {
+        KALDI_ERR << "In edits-config, expected learning-rate-factor to be set in line: "
+                  << config_line.WholeLine();
+      }
+      // Note: the learning_rate_factor_  defined in the component
```
Comment (danpovey) [inline src/nnet3/nnet-utils.cc]: please remove this comment, it's not suitable in this context.

**Comment by pegahgh** [comment]
@danpovey This PR is old and you can remove it. I created and updated new transfer learning PR.

**Pull request closed without merging**